fieldpath/element.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldpath
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kubernetes-sigs/structured-merge-diff/value"
+)
+
+// PathElement describes how to select a child field given a containing object.
+type PathElement struct {
```

Comment:
I'm thinking that this is going to be part of the API, because the metadata field is going to look like ... `ClaimedField map[workflowID]Set`, with `type workflowID string`, then all of the structs defined in this PR would end-up being part of the API. Does it mean we should do something specific for this union? And then all the optional fields should have `// +optional` and `omitempty`, etc

Reply:
Yes, the api type will correspond to this (not sure it's wise to use this directly).